Add quickfixes to some warnings and errors

lrytz opened the pull request:
Looked through Parser, Scanner, Namer, Typer and RefChecks.

Some ad hocery in
  - finding positions, for example from a synthetic accessor DefDef that has only an offset position to get the end of the name where to add an explicit return type
  - identifying qualifiers of a selection that require parentheses (`a.toDouble` vs `(a + a).toDouble`)

Hopefully acceptable for the use case.

lrytz committed:
Add quick fixes to more warnings and errors

lrytz commented:
Would be cool `@deprecated("use other instead")` could make it to a quick fix somehow

https://github.com/scala/scala/blob/v2.13.11/src/library/scala/Predef.scala#L351-L352

lrytz commented:
> Who is the intended user? Someone upgrading a project with a hundred annoying warnings?

This is mainly intended for IDEs. The path via `-quickfix` to apply changes in batches has probably very limited use, we'll see what feedback we get. Basically when a new warning becomes available or is being opted in.

> Rewrite makes unintentionally dangerous things like lossy conversion too easy to unintentionally ignore.

I thik fixes provided by the compiler should always maintain semantics. Let's say we add a new useful warning in a minor release, like for `val d: Double = someInt / 2`. In a large codebase, one might want to apply the semantics-preserving rewrite to all existing cases to allow enabling the warning (fatally) for new code being written.

> The feature I liked on my dotty unused imports branch was rewriting the import, which is always benign.

Metals and IntelliJ have organize imports - what would be the difference?

SethTisue commented:
> to apply changes in batches has probably very limited use 

Perhaps it's more useful than it first seems. One always has the option of letting the compiler apply the update everywhere, but then checking each one by hand/eye (e.g. with `git add -p` or `git checkout -p`) before committing anything.

som-snytt commented:
The other day, I considered that it would be nice to externalize both messages and code actions. If only we had an `errno` or `ErrorID`. I will lgtm asap after I try it out and kick a tire.

som-snytt commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -1029,11 +1028,17 @@ self =>
+        val msg = "type application is not allowed for infix operators"
+        migrationWarning(offset, msg, "2.13.11",
+          runReporting.codeAction("use selection", pos, fix, msg))
+      }
```
It seems like a lot of ceremony in the compiler, but the result is neat. I opened a ticket to let the REPL show the edit.

In this case, there is an unnecessary set of parens, a minor nitpick. ((This is a common style from the dotless era.))

```
scala> List(42) map[String] (_.toString)
                ^
       error: type application is not allowed for infix operators
       Scala 3 migration messages are errors under -Xsource:3. Use -Wconf / @nowarn to filter them or add -Xmigration to demote them to warnings.
       Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=scala3-migration [CodeAction(use selection,Some(type application is not allowed for infix operators),List(TextEdit(RangePosition(<console>, 0, 9, 33),List(42).map[String]((_.toString)))))]
```

som-snytt commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -748,12 +742,18 @@ self =>
+        val msg = s"Wrap `${in.name}` in backticks to use it as an identifier, it will become a keyword in Scala 3."
+        deprecationWarning(in.offset, msg, "2.13.7",
+          runReporting.codeAction("add backticks", r2p(in.offset, in.offset, in.offset + in.name.length), s"`${in.name}`", msg, expected = Some((in.name.toString, unit))))
+      }
```
I recently noticed on another PR that the advice is not supplied in the common case:
```
scala> val yield = 42
           ^
       error: illegal start of simple pattern []

scala> val enum = 42
           ^
       warning: Wrap `enum` in backticks to use it as an identifier, it will become a keyword in Scala 3.
```

som-snytt commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -329,7 +323,7 @@ self =>
     }
     val treeBuilder = new ParserTreeBuilder
-    import treeBuilder.{global => _, unit => _, source => _, fresh => _, _}
+    import treeBuilder.{fresh => _, global => _, source => _, unit => _, _}
```
apparently, it used to be in scope order...

som-snytt commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -2095,17 +2104,15 @@ self =>
           hardMigrationWarning(in.offset, msg("deprecated", without), msg("unsupported", without), "2.10.0", actions)
+        } else {
+          val m = msg("unsupported", "just remove `val`")
+          syntaxError(in.offset, m, actions(m))
```
I'm not sure why the messages differ, but it's always been confusing to me. (And I'm no longer a beginner.) Like, is there a hidden message in "bind the value without val"? Some kind of dark magic incantation? This is an example where it is so much easier to just show me the fixed code! than to explain with words.

som-snytt commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -3035,7 +3046,11 @@ self =>
+            val o = nameOffset + name.decode.length
+            runReporting.codeAction("remove ()", r2p(o, o, o + 2), "", msg, expected = Some(("()", unit)))
+          }
+          def warnNilary() = hardMigrationWarning(nameOffset, unaryMsg("deprecated"), unaryMsg("unsupported"), "2.13.4", action)
```
This is the edge of an edge case, but also an example where the code (to supply actions) is finicky.

The following doesn't get an action:
`def unary_~( /* TODO pass context? */ ) = 42`
or less weird
`def unary_~ ( ) = 42`

Consider the style `if( ! bool ) ???` to demonstrate that some coders have special preferences.

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala:
```diff
@@ -196,8 +196,14 @@ trait ContextErrors extends splain.SplainErrors {
+        if (currentUnit.sourceAt(namePos) == tree.symbol.decodedName) runReporting.codeAction("insert explicit type", namePos.focusEnd, s": $inferred", msg)
+        else Nil
+      if (currentRun.isScala3) cx.warning(tree.pos, msg, WarningCategory.Scala3Migration, action)
+      else cx.warning(tree.pos, msg, WarningCategory.OtherImplicitType, action)
```
There is the recent `namePos` attachment; that was so that diagnostics would cover just the name. Oh in fact that's why I just looking at `val yield`, I didn't bother with `val x,y,z = 42`. I did not consider synthetics at all.

lrytz replied:
I even remembered `namePos`. Indeed it's not there on synthetic trees like accessor `DefDef`s. I should probably have tried harder to fix that instead of working around it...

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -1155,7 +1155,12 @@ trait Namers extends MethodSynthesis {
             val help = if (pts != leg) s" instead of $leg" else ""
-            runReporting.warning(tree.pos, s"under -Xsource:3, inferred $pts$help", WarningCategory.Scala3Migration, tree.symbol)
+            val msg = s"under -Xsource:3, inferred $pts$help"
+            val namePos = tree.pos.focus.withEnd(tree.pos.point + tree.symbol.decodedName.length)
```
This looks very `namePos`-like.

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -445,22 +446,33 @@ abstract class RefChecks extends Transform {
-                if (!exempt())
+                if (!exempt()) {
+                  val msg = "method without a parameter list overrides a method with a single empty one"
+                  val namePos = member.pos.focus.withEnd(member.pos.point + member.decodedName.length)
```
I'll give one of these nameposses a try locally.

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1142,15 +1137,20 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
-                    context.warning(tree.pos, s"integral division is implicitly converted (widened) to floating point. Add an explicit `.to${ptSym.name}`.", WarningCategory.LintIntDivToFloat)
+                    val msg = s"integral division is implicitly converted (widened) to floating point. Add an explicit `.to${ptSym.name}`."
+                    context.warning(tree.pos, msg, WarningCategory.LintIntDivToFloat,
+                      runReporting.codeAction("add conversion", tree.pos, s"(${currentUnit.sourceAt(tree.pos)}).to${ptSym.name}", msg))
```
This is the one where we make it (too) easy to make a bad choice. If they really want it, they can do it the old-fashioned way. (Tell the intern to edit all the conversions.)

som-snytt commented on test/junit/scala/tools/nsc/reporters/AbstractCodeActionTest.scala:
```diff
@@ -91,9 +92,90 @@ abstract class AbstractCodeActionTest extends BytecodeTesting {
+  }
+
+  @Test
+  def constructorProcedureSyntax(): Unit = {
```
Just to note that linting that the unit-valued method has parens for a param list is turned off for tests.

That makes sense as a policy because test methods are never called explicitly; though it would be even finer if only test-annotated methods were ignored.

som-snytt commented on test/junit/scala/tools/nsc/reporters/CodeActionTest.scala:
```diff
@@ -5,5 +5,5 @@ import org.junit.runners.JUnit4
 @RunWith(classOf[JUnit4])
 class CodeActionTest extends AbstractCodeActionTest {
-  override def compilerArgs: String = "-Ystop-after:typer -Yvalidate-pos:typer -Yrangepos -deprecation"
+  override def compilerArgs: String = "-Ystop-after:refchecks -deprecation -Xlint"
```
Xlint enables deprecation in case that helps instead of annoying.

som-snytt commented on test/junit/scala/tools/nsc/reporters/CodeActionXsource3Test.scala:
```diff
@@ -36,4 +36,31 @@ class CodeActionXsource3Test extends AbstractCodeActionTest {
+  @Test
+  def infixTypeArg(): Unit = {
+    assertCodeSuggestion("class C { List apply[Int] 1 }", "class C { List.apply[Int](1) }")
+    assertCodeSuggestion("class C { List apply 1 map[Int] identity }", "class C { (List apply 1).map[Int](identity) }")
```
my previous example was `List(42) map[Int] (_ + 1)` type.

som-snytt reviewed (approved):
OK, I approve! of this tedious detail work. Not sure if the follow-up namePos will solve all the "figure out the name pos" issues.

lrytz commented:
Travis spec build failed. Maybe transient, myabe a key expired?

```
3.22s$ sudo apt-get update
Hit:2 http://security.ubuntu.com/ubuntu bionic-security InRelease
Hit:3 http://apt.postgresql.org/pub/repos/apt bionic-pgdg InRelease
Get:1 http://package.perforce.com/apt/ubuntu bionic InRelease [3,655 B]
Hit:4 http://archive.ubuntu.com/ubuntu bionic InRelease
Hit:5 http://archive.ubuntu.com/ubuntu bionic-updates InRelease
Hit:6 https://public.dhe.ibm.com/software/server/POWER/Linux/toolchain/at/ubuntu bionic InRelease
Hit:7 http://archive.ubuntu.com/ubuntu bionic-backports InRelease
Err:1 http://package.perforce.com/apt/ubuntu bionic InRelease
  The following signatures were invalid: EXPKEYSIG 7123CB760FF18869 Perforce Software (Package Signing) <support+packaging@perforce.com>
Reading package lists... Done
W: GPG error: http://package.perforce.com/apt/ubuntu bionic InRelease: The following signatures were invalid: EXPKEYSIG 7123CB760FF18869 Perforce Software (Package Signing) <support+packaging@perforce.com>
E: The repository 'http://package.perforce.com/apt/ubuntu bionic InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
The command "sudo apt-get update" failed and exited with 100 during .
```

SethTisue commented:
@lrytz needs rebase

lrytz commented:
@SethTisue @som-snytt I'm unsure about the new `[quick fix available]` tag in warnings / errors added in https://github.com/scala/scala/pull/10482, it seems noisy. See the last commit in this PR. Should I revert it, put it behind a flag, replace it with `💡` emoji, ...?

lrytz committed:
move `[quick fix available]` to the end of the message

allow skipping it with `-quickfix:silent`

make more error quickfixes available to `-quickfix` rewrite by
using `runReporting.error` instead of `reporter.error`.

som-snytt commented:
I don't like that the `[qfa]` notice is verbose and pushes text right, so it makes messages harder to scan.

I wondered whether it would be OK as trailing text. `warning: something [qfa]`.

If we had a word for `fixable warning: something`.

Maybe we'll need to upgrade to dotty-style message format. Or rust-style, just follow the message with the correction, annotated as a quickfix.

lrytz commented:
Pushing it into the label (`fixable warning:`) needs to be done in the reporter. `ConsoleReporter` for command-line scalac, but then also the `xsbti.Reporter`, which is in sbt's codebase.

Adding it at the end sounds reasonable. Once you know what it means it would be nice to have something short like `[qf]` or 💡...

There are some multi-line messages, I guess that's OK

```

scala> 'hi
       ^
       warning: symbol literal is deprecated; use Symbol("hi") instead [quick fix available]

scala> println
       ^
       warning: Auto-application to `()` is deprecated. Supply the empty argument list `()` explicitly to invoke method println,
       or remove the empty argument list from its definition (Java-defined methods are exempt).
       In Scala 3, an unapplied method like this will be eta-expanded into a function. [quick fix available]

scala> List(1) map[Int] identity
               ^
       error: type application is not allowed for infix operators [quick fix available]
       Scala 3 migration messages are errors under -Xsource:3. Use -Wconf / @nowarn to filter them or add -Xmigration to demote them to warnings.
       Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=scala3-migration
```

I pushed that. Also added support for `-quickfix:silent` to omit the `[quick fix available]` tag in messages.

lrytz commented:
@SethTisue LGTY?

som-snytt commented:
still lgtm or leb (even better). Trailing `[qfa]` is much tidier. I look forward to bulb emoji someday. No doubt after consuming a small amount of kibble we may have other ideas or feedback.

SethTisue commented:
Well, I'm tempted to just hit "merge". But how about `[quickfixable]`? That saves some space versus `[quick fix available]`, but without being too cryptic like `[qfa]`.

lrytz commented:
> But how about `[quickfixable]`?

I agree shorter is better. If you native speakers think it's fine then it's certainly fine with me.

SethTisue committed:
replace "quick fix available" with "quickfixable"

som-snytt commented:
I think it would sound fine in German but not in English.

Further bikeshedding: emit the filter:
```
error: type application is not allowed for infix operators [-quickfix:cat=scala3-migration]
```
or whatever the syntax is. (See, I don't even remember what to do. I guess in an IDE, you just click a widget.)

lrytz commented:
It seems "quickfixable" is googleable.

SethTisue commented:
community build regression

https://scala-ci.typesafe.com/job/scala-2.13.x-jdk17-integrate-community-build/1327/artifact/logs/ammonite-build.log

```
[ammonite] [error] ## Exception when compiling 8 sources to /home/jenkins/workspace/scala-2.13.x-jdk17-integrate-community-build/target-0.9.20/project-builds/ammonite-5906d4f5e753635e9ef3f1b47a028f37c325b112/amm/target/scala-2.13/classes
[ammonite] [error] java.lang.UnsupportedOperationException: Position.point on NoPosition
[ammonite] [error] scala.reflect.internal.util.Position.fail(Position.scala:24)
[ammonite] [error] scala.reflect.internal.util.UndefinedPosition.point(Position.scala:102)
[ammonite] [error] scala.reflect.internal.util.UndefinedPosition.point(Position.scala:97)
[ammonite] [error] scala.tools.nsc.typechecker.RefChecks$RefCheckTransformer.checkOverride$1(RefChecks.scala:467)
```

I'm digging further...

som-snytt commented:
> I'll give one of these nameposses a try locally.

That was a rash promise.

SethTisue commented:
Okay, so the underlying bug was there even in 2.13.11, and I've reported it https://github.com/scala/bug/issues/12851

And then this PR turned that bug into a crasher.

I think addressing 12851 can wait for 2.13.13.

For 2.13.12, I will submit a PR that just makes this not crash the compiler.

som-snytt commented:
Thanks, I will try that later! more rash promises.
